### PR TITLE
http parser implementation based on http-parser 

### PR DESCRIPTION
**Pull request by fafhrd91**
added c-based http parser, cython code is based on https://github.com/MagicStack/httptools

it gives ~15-25% increase

### CODE REVIEW

File: aiohttp/_parser.pyx [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+        upgrade = bool(self._cparser.upgrade)
+        chunked = bool(self._cparser.flags & cparser.F_CHUNKED)
+
+        headers = CIMultiDict(self._headers)
+
+        if upgrade or self._cparser.method == 5: # cparser.CONNECT:
+            self._upgraded = True
+
+        encoding = None
+        enc = headers.get(hdrs.CONTENT_ENCODING)
```
Comment (kxepal) [inline aiohttp/_parser.pyx]: Is there any need to have two variables to reference on content encoding?

File: aiohttp/_parser.pyx [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+
+    def http_version(self):
+        cdef cparser.http_parser* parser = self._cparser
+
+        if parser.http_major == 1:
+            if parser.http_minor == 0:
+                return HttpVersion10
+            elif parser.http_minor == 1:
+                return HttpVersion11
+
```
Comment (kxepal) [inline aiohttp/_parser.pyx:194]: Should we crash if version will be something weird?

File: aiohttp/_parser.pyx [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+            if self._last_error is not None:
+                ex.__context__ = self._last_error
+                self._last_error = None
+            raise ex
+
+        if self._messages:
+            messages = self._messages
+            self._messages = []
+        else:
+            messages = ()
```
Comment (kxepal) [inline aiohttp/_parser.pyx:230]: `self._messages` are list, but here messages becomes a tuple.

File: aiohttp/_parser.pyx [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+
+
+cdef class HttpRequestParser(HttpParser):
+
+    def __init__(self, protocol, loop,
+                 size_t max_line_size=8190, size_t max_headers=32768,
+                 size_t max_field_size=8190):
+         self._init(cparser.HTTP_REQUEST, protocol, loop,
+                    max_line_size, max_headers, max_field_size)
+        #self._proto_on_url = getattr(protocol, 'on_url', None)
```
Comment (kxepal) [inline aiohttp/_parser.pyx:245]: May be drop this?

File: aiohttp/_parser.pyx [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+
+    pyparser._headers = []
+    return 0
+
+
+cdef int cb_on_url(cparser.http_parser* parser,
+                   const char *at, size_t length) except -1:
+    cdef HttpParser pyparser = <HttpParser>parser.data
+    try:
+        #pyparser.url = _parse_url(at[:length], length)
```
Comment (kxepal) [inline aiohttp/_parser.pyx]: May be remove this?

File: aiohttp/_parser.pyx [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+        pyparser._last_error = ex
+        return -1
+    else:
+        return 0
+
+
+cdef int cb_on_chunk_header(cparser.http_parser* parser) except -1:
+    cdef HttpParser pyparser = <HttpParser>parser.data
+    try:
+        pass
```
Comment (kxepal) [inline aiohttp/_parser.pyx]: Something TBD here?

File: aiohttp/_parser.pyx [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+        pyparser._last_error = ex
+        return -1
+    else:
+        return 0
+
+
+cdef int cb_on_chunk_complete(cparser.http_parser* parser) except -1:
+    cdef HttpParser pyparser = <HttpParser>parser.data
+    try:
+        pass
```
Comment (kxepal) [inline aiohttp/_parser.pyx]: Something TBD here?

File: aiohttp/_parser.pyx [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+        cls = BadHttpMessage
+
+    return cls(desc.decode('latin-1'))
+
+
+@cython.freelist(250)
+cdef class URL:
+    cdef readonly str schema
+    cdef readonly str host
+    cdef readonly object port
```
Comment (kxepal) [inline aiohttp/_parser.pyx]: Why port is an object?

File: aiohttp/server.py [hunk 9]
```diff
@@ -248,11 +260,78 @@ def set_parser(self, parser):
...
+                self._closing = True
+                self._request_handlers.append(
+                    ensure_future(
+                        self.handle_error(
+                            400, None,
+                            None, exc, exc.headers, exc.message),
+                        loop=self._loop))
+                return
+            except Exception as exc:
+                print('exc')
```
Comment (kxepal) [inline aiohttp/server.py]: Debug print.

File: aiohttp/server.py [hunk 10]
```diff
@@ -248,11 +260,78 @@ def set_parser(self, parser):
...
+                self._message_tail = tail
+
+        # no parser, just store
+        elif self._payload_parser is None and self._upgrade:
+            if data:
+                self._message_tail += data
+
+        # feed payload
+        else:
+            if data:
```
Comment (kxepal) [inline aiohttp/server.py]: `elif data`?

File: tests/test_proxy_functional.py [hunk 11]
```diff
@@ -267,9 +275,9 @@ def request(pid):
     yield from sess.close()
 
 
-@pytest.mark.xfail
+# @pytest.mark.xfail
```
Comment (kxepal) [inline tests/test_proxy_functional.py:279]: May be remove all these `# pytest.mark.xfail`?

File: aiohttp/_parser.pyx (hunk 3 above)
Comment (fafhrd91) [reply aiohttp/_parser.pyx:230]: i do not want to convert messages into tuple. doesnt really matter

File: aiohttp/_parser.pyx (hunk 4 above)
Comment (fafhrd91) [reply aiohttp/_parser.pyx:245]: i've not decided on url, http-parser provides url parsing, should give some more performance, but it requires yarl integration

File: aiohttp/_parser.pyx (hunk 5 above)
Comment (fafhrd91) [reply aiohttp/_parser.pyx]: same as previous

File: aiohttp/_parser.pyx (hunk 6 above)
Comment (fafhrd91) [reply aiohttp/_parser.pyx]: i have not decided :)

also i am not sure about 'trailer'

File: aiohttp/_parser.pyx (hunk 7 above)
Comment (fafhrd91) [reply aiohttp/_parser.pyx]: same

File: aiohttp/_parser.pyx (hunk 8 above)
Comment (fafhrd91) [reply aiohttp/_parser.pyx]: i have no idea :) 
url handling needs some more work

File: aiohttp/server.py (hunk 10 above)
Comment (fafhrd91) [reply aiohttp/server.py]: good

File: tests/test_proxy_functional.py (hunk 11 above)
Comment (fafhrd91) [reply tests/test_proxy_functional.py:279]: i do  not understand those tests, i need andrew help

**Comment by fafhrd91** [comment]
added http-parser's url parsing functionality.

on my mac i get ~14.3k rps, ~7k was on 1.3

**Comment by fafhrd91** [comment]
1.2 is slightly faster than 1.3, ~7.3k rps

**Comment by fafhrd91** [comment]
with all latest changes i could get ~16.2rps

interestingly, `__new__` method adds huge overhead.

**Comment by fafhrd91** [comment]
cool, i didn't know about `purest_addoption`

**Comment by fafhrd91** [comment]
@samuelcolvin added `--fast` option

**Comment by samuelcolvin** [comment]
great!

File: aiohttp/test_utils.py [hunk 12]
```diff
@@ -466,18 +459,21 @@ def setup_test_loop(loop_factory=asyncio.new_event_loop):
     return loop
 
 
-def teardown_test_loop(loop):
+def teardown_test_loop(loop, fast=None):
     """Teardown and cleanup an event_loop created
     by setup_test_loop.
 
     """
+    if fast is None:
```
Comment (samuelcolvin) [inline aiohttp/test_utils.py]: I think this should be

```python
fast |= TESTS_FAST
```

that way if you're testing with pytest and have `AIOHTTP_TESTS_FAST=True` but don't supply the `--fast` argument you get fast tests.

With current setup `AIOHTTP_TESTS_FAST` has no effect with pytest.

Comment (fafhrd91) [reply aiohttp/test_utils.py]: I think `AIOHTTP_TESTS_FAST` is not needed

Comment (samuelcolvin) [reply aiohttp/test_utils.py]: fine by me.

**Comment by fafhrd91** [comment]
i modified gunicorn worker to disable access log if access log is not configured

on my mac i got ~60k rps

t_srv.py:
```python 
async def intro(request):
    resp = Response(status=200)
    return resp

def init():
    app = Application()
    app.router.add_get('/', intro)
    return app

app = init()
```

gunicorn command:
```
>>> python -OO -m gunicorn.app.wsgiapp --worker-class aiohttp.worker.GunicornUVLoopWebWorker t_srv:app --workers 6 -b 127.0.0.1:8080
```

and wrk command:
```
wrk -t12 -c400 -d10s http://127.0.0.1:8080/ -s pipeline.lua --latency -- / 16
```

**Comment by fafhrd91** [comment]
merged to master

@kxepal if you want to modify cython code, just modify on master

**Pull request merged**